provider/ecs.go

```diff
@@ -0,0 +1,478 @@
+	service             *ecs.Service
+	task                *ecs.Task
+	taskDefinition      *ecs.TaskDefinition
+	container           *ecs.Container
+	containerDefinition *ecs.ContainerDefinition
+	machine             *ec2.Instance
+}
+
+type awsClient struct {
+	ecs *ecs.ECS
+	ec2 *ec2.EC2
+}
+
+func (provider *ECS) createClient() (*awsClient, error) {
+	sess := session.New()
+	ec2meta := ec2metadata.New(sess)
+	if provider.Region == "" {
+		log.Printf("No EC2 region provided, querying instance metadata endpoint...")
+		identity, err := ec2meta.GetInstanceIdentityDocument()
+		if err != nil {
+			return nil, err
+		}
+		provider.Region = identity.Region
+	}
+
+	cfg := &aws.Config{
+		Region: &provider.Region,
+		Credentials: credentials.NewChainCredentials(
+			[]credentials.Provider{
+				&credentials.StaticProvider{
+					Value: credentials.Value{
+						AccessKeyID:     provider.AccessKeyID,
+						SecretAccessKey: provider.SecretAccessKey,
+					},
+				},
+				&credentials.EnvProvider{},
+				&credentials.SharedCredentialsProvider{},
+				defaults.RemoteCredProvider(*(defaults.Config()), defaults.Handlers()),
+			}),
+	}
+
+	return &awsClient{
+		ecs.New(sess, cfg),
+		ec2.New(sess, cfg),
+	}, nil
+}
+
+// Provide allows the provider to provide configurations to traefik
+// using the given configuration channel.
+func (provider *ECS) Provide(configurationChan chan<- types.ConfigMessage, pool *safe.Pool, constraints types.Constraints) error {
+
+	provider.Constraints = append(provider.Constraints, constraints...)
+
+	pool.Go(func(stop chan bool) {
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			select {
+			case <-stop:
+				cancel()
+			}
+		}()
+
+		operation := func() error {
+			select {
+			case <-ctx.Done():
+				if ctx.Err() == context.Canceled {
+					return nil
+				}
+				return ctx.Err()
+			default:
+			}
+
+			aws, err := provider.createClient()
+			if err != nil {
+				return err
+			}
+
+			configuration, err := provider.loadECSConfig(ctx, aws)
+			if err != nil {
+				return err
+			}
+
+			configurationChan <- types.ConfigMessage{
+				ProviderName:  "ecs",
+				Configuration: configuration,
+			}
+
+			if provider.Watch {
+				reload := time.NewTicker(time.Second * time.Duration(provider.RefreshSeconds))
+				defer reload.Stop()
+				for {
+					select {
+					case <-reload.C:
+						configuration, err := provider.loadECSConfig(ctx, aws)
+						if err != nil {
+							return err
+						}
+
+						configurationChan <- types.ConfigMessage{
+							ProviderName:  "ecs",
+							Configuration: configuration,
+						}
+					case <-ctx.Done():
+						if ctx.Err() == context.Canceled {
+							return nil
+						}
+						return ctx.Err()
+					}
+				}
+			}
+
+			return nil
+		}
+
+		notify := func(err error, time time.Duration) {
+			log.Errorf("ECS connection error %+v, retrying in %s", err, time)
+		}
+		err := backoff.RetryNotify(operation, job.NewBackOff(backoff.NewExponentialBackOff()), notify)
```

Comment:
Could you wrap it with `safe.OperationWithRecover(operation)`,  in case of a panic in the goroutine?